tiledb/sm/query/ast/query_ast.cc

```diff
@@ -0,0 +1,130 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Implements Query AST classes.
+ */
+
+#include "query_ast.h"
+
+namespace tiledb {
+namespace sm {
+
+tdb_unique_ptr<ASTNode> ast_combine(
+    const tdb_unique_ptr<ASTNode>& lhs,
+    const tdb_unique_ptr<ASTNode>& rhs,
+    QueryConditionCombinationOp combination_op) {
+  // AST Construction
+  std::vector<tdb_unique_ptr<ASTNode>> ast_nodes;
+  if (lhs->get_tag() == ASTNodeTag::VAL) {
+    if (rhs->get_tag() == ASTNodeTag::VAL) {
+      ast_nodes.push_back(lhs->clone());
+      ast_nodes.push_back(rhs->clone());
+    } else {
+      // rhs is expression, lhs is value
+      auto rhs_tree_expr = dynamic_cast<ASTNodeExpr*>(rhs.get());
+      ast_nodes.push_back(lhs->clone());
+      for (const auto& elem : rhs_tree_expr->nodes_) {
+        ast_nodes.push_back(elem->clone());
+      }
+    }
+  } else if (rhs->get_tag() == ASTNodeTag::VAL) {
+    // lhs is expression, rhs is value
+    auto lhs_tree_expr = dynamic_cast<ASTNodeExpr*>(lhs.get());
+    for (const auto& elem : lhs_tree_expr->nodes_) {
+      ast_nodes.push_back(elem->clone());
+    }
+    ast_nodes.push_back(rhs->clone());
+  } else {
+    // Both trees are expression trees
+    auto lhs_tree_expr = dynamic_cast<ASTNodeExpr*>(lhs.get());
+    auto rhs_tree_expr = dynamic_cast<ASTNodeExpr*>(rhs.get());
+
+    if (combination_op == lhs_tree_expr->combination_op_ &&
+        lhs_tree_expr->combination_op_ == rhs_tree_expr->combination_op_) {
+      // same op
+      for (const auto& elem : lhs_tree_expr->nodes_) {
+        ast_nodes.push_back(elem->clone());
+      }
+      for (const auto& elem : rhs_tree_expr->nodes_) {
+        ast_nodes.push_back(elem->clone());
+      }
+    } else if (combination_op == lhs_tree_expr->combination_op_) {
+      for (const auto& elem : lhs_tree_expr->nodes_) {
+        ast_nodes.push_back(elem->clone());
+      }
+      ast_nodes.push_back(rhs->clone());
+    } else if (combination_op == rhs_tree_expr->combination_op_) {
+      ast_nodes.push_back(lhs->clone());
+      for (const auto& elem : rhs_tree_expr->nodes_) {
+        ast_nodes.push_back(elem->clone());
+      }
+    } else {
+      ast_nodes.push_back(lhs->clone());
+      ast_nodes.push_back(rhs->clone());
+    }
+  }
+  return tdb_unique_ptr<ASTNode>(
+      tdb_new(ASTNodeExpr, std::move(ast_nodes), combination_op));
+}
+
+void ast_get_field_names(
+    std::unordered_set<std::string>& field_name_set,
+    const tdb_unique_ptr<ASTNode>& node) {
+  if (!node)
+    return;
+  if (node->get_tag() == ASTNodeTag::VAL) {
+    auto val_ptr = dynamic_cast<ASTNodeVal*>(node.get());
+    field_name_set.insert(val_ptr->field_name_);
+  } else {
+    auto expr_ptr = dynamic_cast<ASTNodeExpr*>(node.get());
```

Comment:
Defensive programming:  Should explicitly still check with an `else if` that the node is an `ASTNodeTag::EXPR` instead of assuming if it isn't a `VAL` it's an `ESPR`.  Final else can then be an `assert(false)`.  Similarly in the ifs above.

Reply:
Resolved locally